### PR TITLE
doc/website/blog: openapi generator

### PR DESCRIPTION
**Pull request by masseelch**
@rotemtam @a8m

### CODE REVIEW

File: doc/website/blog/2021-09-10-openapi-generator.md [hunk 1]
```diff
@@ -0,0 +1,329 @@
...
+```shell
+go install github.com/masseelch/elk
+```
+
+`elk` uses the Ent [Extension API](https://entgo.io/docs/extensions) to integrate with Ent’s code-generation. This
+requires that we use the `entc` (ent codegen) package as
+described [here](https://entgo.io/docs/code-gen#use-entc-as-a-package). Follow the next two steps to enable it and to
+configure Ent to work with the `elk` extension:
+
+1. Create a new Go file named `ent/entc.go` and paste the following content:
```
Comment (a8m) [inline doc/website/blog/2021-09-10-openapi-generator.md:31]:
```suggestion
1\. Create a new Go file named `ent/entc.go` and paste the following content:
```
before 
image

After:
image

File: doc/website/blog/2021-09-10-openapi-generator.md [hunk 2]
```diff
@@ -0,0 +1,329 @@
...
+		log.Fatalf("creating elk extension: %v", err)
+	}
+	err = entc.Generate("./schema", &gen.Config{}, entc.Extensions(ex))
+	if err != nil {
+		log.Fatalf("running ent codegen: %v", err)
+	}
+}
+```
+
+2. Edit the `ent/generate.go` file to execute the `ent/entc.go` file:
```
Comment (a8m) [inline doc/website/blog/2021-09-10-openapi-generator.md:60]:
```suggestion
2\. Edit the `ent/generate.go` file to execute the `ent/entc.go` file:
```

File: doc/website/blog/2021-09-10-openapi-generator.md [hunk 3]
```diff
@@ -0,0 +1,329 @@
...
+}
+```
+
+Now, let's generate the Ent code and the OAS file.
+
+```shell
+go generate ./...
+```
+
+In addition to the files Ent normally generates, another file names `openapi.json` has been created. Copy its contents
```
Comment (a8m) [inline doc/website/blog/2021-09-10-openapi-generator.md:183]:
```suggestion
In addition to the files Ent normally generates, another file named `openapi.json` has been created. Copy its contents
```

File: doc/website/blog/2021-09-10-openapi-generator.md [hunk 4]
```diff
@@ -0,0 +1,329 @@
...
+and paste them into the [Swagger Editor](https://editor.swagger.io/), You should see three groups: Compartment, Content
+and Fridge. If you happen to open up the POST operation tab in the Fridge group, you see a description of the expected
```
Comment (a8m) [inline doc/website/blog/2021-09-10-openapi-generator.md:185]:
```suggestion
and paste them into the [Swagger Editor](https://editor.swagger.io/), You should see three groups: `Compartment`, `Content`
and `Fridge`. If you happen to open up the POST operation tab in the Fridge group, you see a description of the expected
```

I think groups should be highlighted somehow.

Also, maybe add a screenshot from the Swagger editor?

File: doc/website/blog/2021-09-10-openapi-generator.md [hunk 5]
```diff
@@ -0,0 +1,329 @@
...
+	"entgo.io/ent/entc/gen"
+	"github.com/masseelch/elk"
+)
+
+func main() {
+	ex, err := elk.NewExtension(
+		elk.GenerateSpec(
+			"openapi.json",
+			elk.SpecTitle("Fridge CMS"), // It is a Content-Management-System ...
+			elk.SpecDescription("API to manage fridges and their cooled contents. **ICY!**"), // You can use CommonMark syntax.
```
Comment (a8m) [inline doc/website/blog/2021-09-10-openapi-generator.md:213]:
```suggestion
			elk.SpecTitle("Fridge CMS"),
			// You can use CommonMark syntax.
			elk.SpecDescription("API to manage fridges and their cooled contents. **ICY!**"),
```

The second is hidden by default. Maybe move it atop?

File: doc/website/blog/2021-09-10-openapi-generator.md [hunk 6]
```diff
@@ -0,0 +1,329 @@
...
+operation on the Fridge by using an `elk.SchemaAnnotation`:
+
+```go title="ent/schema/fridge.go"
+// Annotations of the Fridge.
+func (Fridge) Annotations() []schema.Annotation {
+	return []schema.Annotation{
+		elk.DeletePolicy(elk.Exclude),
+	}
+}
+```
```
Comment (a8m) [inline doc/website/blog/2021-09-10-openapi-generator.md:243]: Beautiful! 😍

File: doc/website/blog/2021-09-10-openapi-generator.md [hunk 7]
```diff
@@ -0,0 +1,329 @@
...
+
+And voilà, the DELETE operation is gone. For more information about how `elk`s policies work and what you can do with
+it, have a look at the [godoc](https://pkg.go.dev/github.com/masseelch/elk).
+
+### Extend specification
+
+The one thing I should be interested the most in this example is the current contents of a fridge.You can customize the
```
Comment (a8m) [inline doc/website/blog/2021-09-10-openapi-generator.md:250]:
```suggestion
The one thing I should be interested the most in this example is the current contents of a fridge. You can customize the
```

**Review by a8m — APPROVED** [review]
Fantastic work @masseelch! I appreciate your contribution to Ent and really love what you do. `elk` is super cool 🙏 ❤️

**Comment by a8m** [comment]
cc @giautm, if you want to review 🙏

File: doc/website/blog/2021-09-10-openapi-generator.md (hunk 4 above)
Comment (masseelch) [reply doc/website/blog/2021-09-10-openapi-generator.md:185]: Agree to the highliting, but I think making them bold/strong is fits more than backticks.

I will add some screenshots to the post.

File: doc/website/blog/2021-09-10-openapi-generator.md [hunk 8]
```diff
@@ -0,0 +1,350 @@
...
+	}
+}
+```
+
+```shell
+go run -mod=mod main.go
+```
+
+Our Fridge API is up and running. With the generated OAS file and the Swagger Tooling you can generate any client stub
+you want and speed up the development by several orders of magnitude.
```
Comment (rotemtam) [inline doc/website/blog/2021-09-10-openapi-generator.md:335]:
```suggestion
Our Fridge API server is up and running. With the generated OAS file and the Swagger Tooling you can now generate a client stub
in any supported language and forget about writing a RESTful client ever _ever_ again.
```

File: doc/website/blog/2021-09-10-openapi-generator.md [hunk 9]
```diff
@@ -0,0 +1,350 @@
...
+
+### Wrapping Up
+
+In this post we introduced a new feature of `elk` - OAS generation. This enables you to use the power of Swagger to
+speed up development, have rich documentation and much more.
```
Comment (rotemtam) [inline doc/website/blog/2021-09-10-openapi-generator.md:340]:
```suggestion
In this post we introduced a new feature of `elk` - automatic OpenAPI Specification generation. This feature connects between Ent's code-generation capabilities and OpenAPI/Swagger's rich tooling ecosystem. 
```

**Review by rotemtam — CHANGES_REQUESTED** [review]
@masseelch ,
This is an amazing feature and a beautiful blog post.
Please find my humble suggestions to this excellent work.
I have a gut feeling that this will explode in our community.
Good luck!

File: doc/website/blog/2021-09-10-openapi-generator.md [hunk 10]
```diff
@@ -0,0 +1,350 @@
...
+	"log"
+
+	"entgo.io/ent/entc"
+	"entgo.io/ent/entc/gen"
+	"github.com/masseelch/elk"
+)
+
+func main() {
+	ex, err := elk.NewExtension(
+		elk.GenerateSpec("openapi.json"),
```
[the remark this answers is not in the document]
Comment (a8m) [reply doc/website/blog/2021-09-10-openapi-generator.md:48]: I find this comment unnecessary tbh, as the `GenerateSpec` is self-explanatory and the logic is explained after this code block.

File: doc/website/blog/2021-09-10-openapi-generator.md [hunk 11]
```diff
@@ -0,0 +1,350 @@
...
+The first step on our way to the OAS file is to create an Ent schema graph:
+
+```shell
+go run -mod=mod entgo.io/ent/cmd/ent init Fridge Compartment Content
+```
+
+So, for an example: I have multiple fridges with multiple compartments, and me and my SO want to know its contents at all times. We will
+create a Go server to hold the state and with the generated OAS file you are invited to build a frontend to manage
+fridges and contents in the language or your choice by using the Swagger Codegen. You can find a `generate.go` that uses
+docker to generate a client [here](https://github.com/masseelch/elk/blob/master/internal/openapi/ent/generate.go).
```
[the remark this answers is not in the document]
Comment (a8m) [reply doc/website/blog/2021-09-10-openapi-generator.md:83]: I'd put a space between the 2 generate commands:

```diff
// Generate ent code and swagger file.
//go:generate go run -mod=mod entc.go
+
// Example of how to use the OAS file to create a go client for our server with the swagger-codegen docker image.
// go:generate bash -c "docker run --rm -u \"$(id -u):$(id -g)\" -v ${PWD}/../:/local swaggerapi/swagger-codegen-cli-v3 generate -i /local/ent/openapi.json -l go -o /local/client"
```

File: doc/website/blog/2021-09-10-openapi-generator.md [hunk 12]
```diff
@@ -0,0 +1,350 @@
...
+
+```go
+package ent
+
+//go:generate go run -mod=mod entc.go
+```
+
+With these steps complete, all is set up for generating an OAS file from your schema! If you are new to Ent and want to
+learn more about it, how to connect to different types of databases, run migrations or work with entities, then head
+over to the [Setup Tutorial](https://entgo.io/docs/tutorial-setup/).
```
[the remark this answers is not in the document]
Comment (masseelch) [reply doc/website/blog/2021-09-10-openapi-generator.md:70]: :-)

File: doc/website/blog/2021-09-10-openapi-generator.md [hunk 13]
```diff
@@ -0,0 +1,350 @@
...
+```shell
+go generate ./...
+```
+
+In addition to the files Ent normally generates, another file named `openapi.json` has been created. Copy its contents
+and paste them into the [Swagger Editor](https://editor.swagger.io/). You should see three groups: **Compartment**, **
+Content** and **Fridge**. 
+
+<div style={{textAlign: 'center'}}>
+  <img alt="This is what Swagger Editor shows" src="https://via.placeholder.com/150" />
```
[the remark this answers is not in the document]
Comment (masseelch) [reply doc/website/blog/2021-09-10-openapi-generator.md:188]: Already sent to @a8m.

File: doc/website/blog/2021-09-10-openapi-generator.md [hunk 14]
```diff
@@ -0,0 +1,350 @@
...
+	"github.com/masseelch/elk"
+)
+
+func main() {
+	ex, err := elk.NewExtension(
+		elk.GenerateSpec(
+			"openapi.json",
+			// It is a Content-Management-System ...
+			elk.SpecTitle("Fridge CMS"), 
+			// You can use CommonMark syntax.
```
[the remark this answers is not in the document]
Comment (masseelch) [reply doc/website/blog/2021-09-10-openapi-generator.md:226]: True. But its a blog post, it wants to teach! I can add a link to the spec :+1:

**Review by rotemtam — APPROVED** [review]

**Pull request merged**